lib/llvmIr.js

```diff
@@ -0,0 +1,127 @@
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+const utils = require('./utils');
+
+class LlvmIrParser {
+    constructor(compilerProps) {
+        // @todo More settings/compiler properties
+
+        this.maxIrLines = 500;
+        if (compilerProps) {
+            this.maxIrLines = compilerProps('maxLinesOfAsm', this.maxIrLines);
+        }
+    }
+
+    trimIrLine(line, filters) {
+        if (!filters.trim) return line;
+        const splat = line.split(/\s+/);
+        if (splat[0] === "") {
+            // An indented line: preserve a two-space indent
+            return "  " + splat.slice(1).join(" ");
+        } else {
+            return splat.join(" ");
+        }
+    }
+
+    processIr(ir, filters) {
+        const result = [];
+        let irLines = utils.splitLines(ir);
+
+        let debugInfo = {};
+
+        // @todo Get source filename ("source_filename = ...").
+
+        irLines.forEach(line => {
+            let source = null;
+            let match;
+
+            if (line.trim() === "") {
+                result.push({text: "", source: null});
+                return;
+            }
+
+            // Non-Meta IR line. Metadata is attached to it using "!dbg !123"
+            match = line.match(/!dbg (!\d+)/);
+            if (match) {
+                result.push({text: this.trimIrLine(line, filters), source: source, scope: match[1]});
+                return;
+            }
+
+            // Metadata Nodes
+            // See: https://llvm.org/docs/LangRef.html#metadata
+            // @todo: Multiline Metadata?
+            // @todo: Named Metadata?
+            match = line.match(/^(!\d+) = (distinct )?!DI([A-Za-z]+)\(([^)]+?)\)/);
+            if (match) {
+                const metaId = match[1];
+                const metaType = match[2];
+                if (!metaType) { return; }
+
+                let options = {};
+                const optionWithValue = match[4].split(',');
+                for (let i = 0; i < optionWithValue.length; ++i) {
+                    const pair = optionWithValue[i].split(':');
+                    options[pair[0].trim()] = pair[1].trim();
+                }
+                options.metaId = metaId;
+
+                if (options.line) {
+                    debugInfo[metaId] = options;
+                }
+            }
+
+            result.push({text: this.trimIrLine(line, filters), source: source});
+        });
+
+        function getLineByMetaId(metaId) {
```

Comment:
this inline function threw me off, for readability I think it would be better if you'd place it outside and store debugInfo[] in 'this'

Reply:
I think `getSourceLineByMetaId` would also be a better name (this function is only used in that one specific place and could also be placed inline in lilne 103; I actually put it in a function for readability :smile:). `debugInfo` should still be a local variable as it is created for each call to `processIr()`.

Reply:
Renamed `llvmIr.js`  to `llvm-ir.js` to stick with conventions. I removed that function and placed it inline.